### ISSUE TITLE
Feature: GANEstimator allow passing of namedtuples

### ISSUE THREAD

**Issue by sleighsoft** [tensorflow/tensorflow#15157]
In the current `GANEstimator` implementation in `train.py` `gan_model(..)` both `real_data` as well as `generator_inputs` is converted to tensors with either `_convert_tensor_or_l_or_d` or `ops.convert_to_tensor`. This prevents the user from using own data structures like `namedtuples` to pass information between the `generator` and `discriminator`.

In the current implementation when passing a `namedtuple` the result will be a `list` with all name information being lost.

I propose to either extend the tensor conversion to exclude namedtuples from them or to remove them entirely.

@joel-shor Do you think that is a good idea? I am currently passing logits as well as sample_id from a dynamic_decoding around and I would like to keep the meaning of these across loss_fn and discriminator_fn.

1. I could remove the conversions entirely. This would introduce breaking changes.
2. I could exclude namedtuples from the conversion.
3. Idk yet??

I would create a PR for any of them if we find a suitable solution.

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by facaiy** [comment]
Hi, I am not familiar with gan estimator, however many function support dict in tensorflow. Could dict solve the problem?

**Comment by joel-shor** [comment]
@facaiy made the point I wanted to make: TFGAN already supports dictionaries. Could you just use the namedtuple `_asdict()` function for the duration of the GANEstimator? I'm also happy to have you add support for nametuples in `_convert_tensor_or_l_or_d`, which I think would also fix your issue?

**Comment by sleighsoft** [comment]
I would still love to see namedtuples for their immutability but I can live for now with a dictionary.
@joel-shor You can decide wether you leave this request open or close it.
